apps/src/authoredHintUtils.js

```diff
@@ -0,0 +1,221 @@
+/**
+ * @overview A helper class for all actions associated with the Authored
+ * Hint viewing and logging system.
+ *
+ * The general pattern for usage is straightforward:
+ *
+ * Every time a user requests to view a new hint, we log information
+ * about that hint as well as the most recent attempt.
+ *
+ * Every time a user makes an "attempt" at the level and send an attempt
+ * report, we save some information about that most recent attempt, and
+ * add that information to all hint views from the previous attempt.
+ *
+ * Finally, when the user navigates to a new page (either by completing
+ * the level and moving on or by leaving the level and coming back
+ * later), we record for all hints the "final" attempt in that session
+ * and post the results to the server.
+ *
+ * Thus, each hint has three attempt records by the end of the
+ * lifecycle:
+ *  1) a "previous" attempt (can be empty)
+ *  2) a "next" attempt
+ *  3) a "final" attempt
+ */
+
+/**
+ * @typedef {Object} HintData
+ * @property {number} script_id
+ * @property {number} level_id
+ * @property {string} hint_id
+ * @property {string} hint_class
+ * @property {string} hint_type
+ *
+ * @typedef {Object} UnfinishedHint
+ * @augments HintData
+ * @property {number} [prev_time]
+ * @property {number} [prev_attempt]
+ * @property {number} [prev_test_result]
+ * @property {number} [prev_activity_id]
+ * @property {number} [prev_level_source_id]
+ *
+ * @typedef {Object} FinishedHint
+ * @augments UnfinishedHint
+ * @property {number} next_time
+ * @property {number} next_attempt
+ * @property {number} next_test_result
+ * @property {number} next_activity_id
+ * @property {number} next_level_source_id
+ *
+ * @typedef {Object} FinalizedHint
+ * @augments FinishedHint
+ * @property {number} final_time
+ * @property {number} final_attempt
+ * @property {number} final_test_result
+ * @property {number} final_activity_id
+ * @property {number} final_level_source_id
+ *
+ * @typedef {Object} AttemptRecord
+ * @property {number} time
+ * @property {number} attempt
+ * @property {number} test_result
+ * @property {number} activity_id
+ * @property {number} level_source_id
```

Comment:
I'm not a big fan of all the snake_case. I realize this is a bit of a unique situation in that these are probably fields that are used in both dashboard (where we use snake_case) and apps (where we use camelCase). My preference, if possible, would be that since we're doing so much processing in JS that we make these camelCase, and then perhaps in dashboard code convert to camel_case if needed. Thoughts?